File: Arrays_And_Hashing/0242-ValidAnagram/0242-hashmap-sol.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        countS= {}
        countT= {}
        # Create two dictionaries, which are like HashMaps in python
        # looping through each string, if character is a key in dictionary, add 1 to its value.
        # if character is not a key in the dictionary, add the key and value 1.

        # compare the dictionaries to check if they are equal.

        # Time complexity should be O(n) for placing keys into the dictionary 
        for i in range(len(s)):
            if s[i] in countS: countS[s[i]] +=1
            else: countS[s[i]] = 1
        
        for i in range(len(t)):
            if t[i] in countT: countT[t[i]] +=1
            else: countT[t[i]] = 1
        if countS == countT: 
            return True
        else: return False

File: Arrays_And_Hashing/0242-ValidAnagram/0242-hashmap-sol.java
class Solution {
    public boolean isAnagram(String s, String t) {
        // Create two hashmaps for string s and t
        // First check if length of s and t is equal. if not, return false as these are not anagrams.
        // now we can just use 1 for loop to check if the hashmaps contains the character.
        // if hash maps don't contain the current character, add it into the hashmap, otherwise increment value of key by 1.
        Map<Character,Integer> map_S = new HashMap<>();
        Map<Character,Integer> map_T = new HashMap<>();
        if(s.length() != t.length()){
            return false;
        }
        for(int i = 0; i < s.length(); i++){
            if(map_S.containsKey(s.charAt(i))){
                map_S.put(s.charAt(i), map_S.get(s.charAt(i))+1);
            }
            else{
                map_S.put(s.charAt(i),1);
            }
            if(map_T.containsKey(t.charAt(i))){
                map_T.put(t.charAt(i), map_T.get(t.charAt(i))+1);
            }
            else{
                map_T.put(t.charAt(i),1);
            }
        }
        if(map_S.equals(map_T)){
            return true;
        }
        else{
            return false;
        }
    }
}